File: frontend/src/services/hospitalService.js
import api from './api';

const hospitalService = {
  getAll: async () => {
    const response = await api.get('/hospitals');
    return response.data;
  },

  getById: async (id) => {
    const response = await api.get(`/hospitals/${id}`);
    return response.data;
  },

  search: async (params) => {
    // params like q, city, type
    const query = new URLSearchParams(params).toString();
    const response = await api.get(`/hospitals/search?${query}`);
    return response.data;
  },

  getNearest: async (lat, lng) => {
    const response = await api.get(`/hospitals/nearest?lat=${lat}&lng=${lng}`);
    return response.data;
  },

  // ─────────── Hospital-owner self-service ───────────
  getMyProfile: async () => {
    const response = await api.get('/hospitals/profile/me');
    return response.data;
  },

  createProfile: async (profileData) => {
    const response = await api.post('/hospitals/profile', profileData);
    return response.data;
  },

  updateProfile: async (profileData) => {
    const response = await api.put('/hospitals/profile/me', profileData);
    return response.data;
  },

  getProfileStatus: async () => {
    const response = await api.get('/hospitals/profile/status');
    return response.data;
  },

  // ─────────── Hospital images ───────────
  addImage: async (hospitalId, imageUrl) => {
    const response = await api.post(`/hospitals/${hospitalId}/images`, { imageUrl });
    return response.data;
  },

  replaceImage: async (imageId, imageUrl) => {
    const response = await api.put(`/hospitals/images/${imageId}`, { imageUrl });
    return response.data;
  },

  deleteImage: async (imageId) => {
    const response = await api.delete(`/hospitals/images/${imageId}`);
    return response.data;
  },

  reorderImages: async (hospitalId, orderedIds) => {
    const response = await api.put(`/hospitals/${hospitalId}/images/reorder`, { orderedIds });
    return response.data;
  },

  // ─────────── Reviews ───────────
  getReviews: async (hospitalId) => {
    const response = await api.get(`/hospitals/${hospitalId}/reviews`);
    return response.data;
  },

  createReview: async (hospitalId, data) => {
    const response = await api.post(`/hospitals/${hospitalId}/reviews`, data);
    return response.data;
  },

  // ─────────── Admin hospital management ───────────
  getAllAdmin: async () => {
    const response = await api.get('/admin/hospitals');
    return response.data;
  },

  updateStatus: async (id, status) => {
    const response = await api.patch(`/admin/hospitals/${id}/status`, { status });
    return response.data;
  },

  getPendingReviews: async () => {
    const response = await api.get('/admin/hospitals/reviews/pending');
    return response.data;
  },

  updateReviewStatus: async (reviewId, status) => {
    const response = await api.put(`/admin/hospitals/reviews/${reviewId}/status`, { status });
    return response.data;
  },

  getAllWithInactive: async () => {
    const response = await api.get('/hospitals', { params: { includeInactive: true } });
    return response.data;
  },

  create: async (data) => {
    const response = await api.post('/hospitals', data);
    return response.data;
  },

  update: async (id, data) => {
    const response = await api.put(`/hospitals/${id}`, data);
    return response.data;
  },

  remove: async (id) => {
    const response = await api.delete(`/hospitals/${id}`);
    return response.data;
  }
};

export default hospitalService;

File: medifind-backend/src/main/java/com/medifind/hospital/service/HospitalService.java
package com.medifind.hospital.service;

import com.medifind.hospital.dto.*;
import com.medifind.hospital.entity.HospitalStatus;

import java.util.List;

/**
 * Business service contract for Hospital operations.
 */
public interface HospitalService {

    /**
     * Create a new hospital record (admin).
     */
    HospitalResponse createHospital(HospitalRequest request);

    /**
     * Retrieve all hospitals.
     *
     * @param includeInactive include deactivated hospitals (admin views)
     */
    List<HospitalResponse> getAllHospitals(boolean includeInactive);

    /**
     * Retrieve a single hospital by its ID.
     */
    HospitalResponse getHospitalById(Long id);

    /**
     * Update an existing hospital record (admin).
     */
    HospitalResponse updateHospital(Long id, HospitalRequest request);

    /**
     * Permanently delete a hospital record (admin). Also removes its images.
     */
    void deleteHospital(Long id);

    // ─────────── Hospital-owner profile operations ───────────

    /**
     * Create the hospital profile owned by an auth user. New profiles start
     * PENDING and only become public once an admin approves them.
     */
    HospitalResponse createProfile(Long userId, HospitalProfileRequest request);

    /**
     * Update the hospital profile owned by an auth user.
     */
    HospitalResponse updateProfile(Long userId, HospitalProfileRequest request);

    /**
     * Fetch the profile owned by an auth user (404 when it does not exist).
     */
    HospitalResponse getProfileByUserId(Long userId);

    /**
     * Profile completion status for the auth flow — "INCOMPLETE" when no
     * profile exists, otherwise the hospital status name.
     */
    String getProfileStatus(Long userId);

    /**
     * List all hospitals (admin view, includes every status).
     */
    List<HospitalResponse> getAllHospitalsAdmin();

    /**
     * Change a hospital's lifecycle status (approve / suspend / activate /
     * reject). Keeps the legacy {@code active} flag in sync with the status.
     */
    HospitalResponse updateStatus(Long id, HospitalStatus status);

    /**
     * Search hospitals by free-text (name / city / address / type / facilities).
     * Only returns hospitals that are publicly visible.
     */
    List<HospitalResponse> searchHospitals(String q, String city, String type);

    // ─────────── Hospital image operations ───────────

    /**
     * Add one image to a hospital. Maximum 10 images per hospital.
     */
    HospitalImageResponse addImage(Long hospitalId, String imageUrl, Long requesterUserId, boolean isAdmin);

    /**
     * Replace an existing image.
     */
    HospitalImageResponse replaceImage(Long imageId, String imageUrl, Long requesterUserId, boolean isAdmin);

    /**
     * Remove an image. When the cover image is removed, the cover falls back
     * to the first remaining image.
     */
    void deleteImage(Long imageId, Long requesterUserId, boolean isAdmin);

    /**
     * Reorder a hospital's images by the given ordered ids.
     */
    List<HospitalImageResponse> reorderImages(Long hospitalId, List<Long> orderedIds, Long requesterUserId, boolean isAdmin);

    /**
     * Delete all images belonging to a hospital (used when deleting the hospital).
     */
    void deleteImagesByHospital(Long hospitalId);
}
